### PR TITLE
RL with vllm-native support (qwen3 converter)

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

Implement/refactor MaxText to vLLM weight conversion into a reusable converter style:

- Introduce `BaseMaxTextToVLLMConverter` as the shared converter class, and isolate the Qwen3-MOE implementation into `torchax_converter/qwen3_moe.py`.
- Add a standalone, config-driven validator module at `maxtext.integration.vllm.torchax_converter.validate_converter` for VM testing and generation-based weight-transfer checks.
- Add a `use_standalone_converter` config flag so RL rollout can explicitly opt into the standalone MaxText to vLLM converter path.
- Update rollout integration to use `MaxTextVllmRollout` with model-specific converter creation instead of importing converter logic from the bench script.
- TP-aware padding to 128x is also added in qwen3 converter `_make_fuse_all`.

# Tests
[requirements](https://paste.googleplex.com/5240747596054528) (tpu-inference/vllm are critical)
[logs](https://paste.googleplex.com/4819117233274880)

Standalone weight sync and decode on a v5p-8 VM:
```
python -m maxtext.integration.vllm.torchax_converter.validate_converter src/maxtext/configs/base.yml model_name=qwen3-30b-a3b tokenizer_type=huggingface tokenizer_path=Qwen/Qwen3-30B-A3B  load_parameters_path=gs://hengtaoguo-maxtext-logs/checkpoints/qwen3-30b-a3b/scanned/2026-01-23-14-00/0/items/0/items run_name=qwen3_converter_validation per_device_batch_size=1 max_prefill_predict_length=8 max_target_length=16 steps=1 scan_layers=true skip_jax_distributed_system=true weight_dtype=bfloat16 attention=dot_product remat_policy=custom decoder_layer_input=offload query_proj=offload key_proj=offload value_proj=offload rollout_tensor_parallelism=4 hbm_utilization_vllm=0.6 async_scheduling=false prompt=Paris\ is hf_access_token=xxx
```

```
Assigning 435 weights to vLLM model took 0.1711 seconds

================================================================================
Generation test after weight transfer:
Rendering prompts: 100%|█| 1/1 [00:00<00:
Processed prompts:   0%| | 0/1 [00:00<?, /home/hengtaoguo_google_com/projects/venv2/lib/python3.12/site-packages/torchax/tensor.py:167: UserWarning: Explicitly requested dtype int64 requested in astype is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/jax-ml/jax#current-gotchas for more.
  res = jax_function(self._elem, *args, **kwargs)
WARNING 04-30 04:23:44 [tuned_block_sizes.py:4368] Couldn`t find tuned sizes for the RPA v3 kernel with ('TPU v5', 16, 'q_bfloat16_kv_bfloat16', 'q_head-8_kv_head-1_head-128', 'max_model_len-16-sw-None')
INFO 04-30 04:23:44 [tuned_block_sizes.py:4389] RPA v3 kernel tuned block sizes for ('TPU v5', 16, 'q_bfloat16_kv_bfloat16', 'q_head-8_kv_head-1_head-128', 'max_model_len-16-sw-None'): bkv_p=1, bq=16
Processed prompts: 100%|█| 1/1 [00:06<00:
[RequestOutput(request_id=0, prompt='Paris is', prompt_token_ids=[59604, 374], encoder_prompt=None, encoder_prompt_token_ids=None, prompt_logprobs=None, outputs=[CompletionOutput(index=0, text=' the capital of France. It is also the city', token_ids=[279, 6722, 315, 9625, 13, 1084, 374, 1083, 279, 3283], routed_experts=None, cumulative_logprob=None, logprobs=None, finish_reason=length, stop_reason=None)], finished=True, metrics=None, lora_request=None, num_cached_tokens=0)]
Generation took 6.9421 seconds
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  raise ValueError(f"Unsupported model for Qwen3 converter validation: {model_name}")
+
+
+def _get_maxtext_model(config):
+  logging.info("Creating model with config: %s", config)
+  model, mesh = model_creation_utils.from_pretrained(
+      config,
+      model_mode=MODEL_MODE_AUTOREGRESSIVE,
+  )
+  with mesh:
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: Do we need to return model from TunixMaxTextAdapter, can we directly return the maxtext_model, mesh

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    tunix_model = TunixMaxTextAdapter(base_model=model)
+    tunix_model.config = _qwen3_model_config(config.model_name)
+  return tunix_model, mesh
+
+
+def _resolve_vllm_model(config) -> str:
+  return getattr(config, "rollout_vllm_model_version", None) or config.tokenizer_path
+
+
+def _resolve_tensor_parallel_size(config) -> int:
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: since this function is supposed use TP for the vllm model, should we just return `rollout_tensor_parallelism`

Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: Good catch! I'm directly assigning `config.rollout_tensor_parallelism` now

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py (hunk 1 above)
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: Done, the helper function now only returns maxtext model since TunixMaxTextAdapter is not used any more.

**Review by khatwanimohit — APPROVED** [review]
LGTM!

File: src/maxtext/integration/vllm/torchax_converter/qwen3_moe.py
Comment (NicoGrande) [inline src/maxtext/integration/vllm/torchax_converter/qwen3_moe.py:1]: Do these transformations apply to both dense and moe qwen3?

File: src/maxtext/integration/vllm/maxtext_vllm_rollout.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+      )
+      del weight, weight_array  # release TPU buffer before pushing back to device
+      # Periodically flush async ops and GC to prevent host RAM accumulation.
+      if i % 16 == 15:
+        jax.effects_barrier()
+        gc.collect()
+    jax.effects_barrier()
+    gc.collect()
+    end_time = time.time()
+    logging.info("MaxTextVllmSampler.update_params: all weights assigned in %.4f seconds", end_time - start_time)
```
Comment (NicoGrande) [inline src/maxtext/integration/vllm/maxtext_vllm_rollout.py:127]: This block of code is your weight sync function right? Could we create a helper function which executes weight sync and then call that instead? Also, it would be nice if you used the `config.reshard_chunk_size` argument to control how often to reshard, gc.collect(), etc.

**Review by NicoGrande — APPROVED** [review]
LGTM!

File: src/maxtext/integration/vllm/torchax_converter/qwen3_moe.py
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/qwen3_moe.py:1]: I have only tested this on the moe models. For dense models, we can set `use_standalone_converter=False` and use the existing mappings.

File: src/maxtext/integration/vllm/maxtext_vllm_rollout.py (hunk 3 above)
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/maxtext_vllm_rollout.py:127]: Great call, let me follow up in the next PR to test with e2e RL runs, with better helper functions and cache clearing schemes.
